crud/select/plan.lua

```diff
@@ -48,49 +48,75 @@ local function get_index_for_condition(space_indexes, space_format, condition)
+local function extract_sharding_key_from_conditions(conditions, sharding_index, space_indexes, fieldno_map)
+    if sharding_index == nil then
         return nil
     end
 
-    if scan_index.id == sharding_index.id then
-        return scan_value
-    end
+    -- If name is both valid index name and field name,
+    -- it is interpreted as index name.
+    local filled_fields = {}
+    for _, condition in ipairs(conditions) do
+        if condition.operator ~= compare_conditions.operators.EQ then
+            goto continue
+        end
 
-    local scan_value_fields_values = {}
-    for i, scan_index_part in ipairs(scan_index.parts) do
-        scan_value_fields_values[scan_index_part.fieldno] = scan_value[i]
-    end
+        local index = space_indexes[condition.operand]
+        if index ~= nil then
+            for i, part in ipairs(index.parts) do
+                filled_fields[part.fieldno] = condition.values[i]
+            end
 
-    -- check that sharding key is included in the scan index fields
-    local sharding_key = {}
-    for _, sharding_key_part in ipairs(sharding_index.parts) do
-        local fieldno = sharding_key_part.fieldno
+            goto continue
+        end
 
-        -- sharding key isn't included in scan key
-        if scan_value_fields_values[fieldno] == nil then
-            return nil
+        local fieldno = fieldno_map[condition.operand]
+        if fieldno == nil then
+            goto continue
         end
+        filled_fields[fieldno] = condition.values[1]
 
-        local field_value = scan_value_fields_values[fieldno]
+        ::continue::
+    end
 
-        -- sharding key contains nil values
-        if field_value == nil then
+    local sharding_key = {}
+    for i, v in ipairs(sharding_index.parts) do
+        if filled_fields[v.fieldno] == nil then
             return nil
         end
 
-        table.insert(sharding_key, field_value)
+        sharding_key[i] = filled_fields[v.fieldno]
     end
 
     return sharding_key
 end
 
+local function get_sharding_key_from_scan_value(scan_value, scan_index, scan_iter, sharding_index)
+    if scan_value == nil then
+        return nil
+    end
+
+    if scan_iter ~= box.index.EQ and scan_iter ~= box.index.REQ then
+        return nil
+    end
+
+    if #scan_value < #sharding_index.parts then
+        return nil
+    end
+
+    if scan_index.id == sharding_index.id then
+        return scan_value
+    end
+
+    return nil
+end
+
 -- We need to construct after_tuple by field_names
 -- because if `fields` option is specified we have after_tuple with partial fields
 -- and these fields are ordered by field_names + primary key + scan key
 -- this order can be differ from order in space format
 -- so we need to cast after_tuple to space format for scrolling tuples on storage
-local function construct_after_tuple_by_fields(space_format, field_names, tuple)
+local function construct_after_tuple_by_fields(fieldno_map, field_names, tuple)
```

Comment:
I would add unit tests for function `construct_after_tuple_by_fields()`. At least, case `with error is not covered.

Reply:
Since `plan.new` now uses `fieldno_map` it would be weird to extract it multiple times inside different functions, thus I adapted `construct_after_tuple_by_fields` to use it. But this function does not related in any way to this PR and I don't understand why it should be covered with unit tests as part of this PR.